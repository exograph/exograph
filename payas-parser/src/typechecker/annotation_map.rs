use std::collections::{BTreeMap, HashMap};
use std::fmt::Debug;

use crate::ast::ast_types::{AstAnnotation, AstAnnotationParams, Untyped};
use crate::typechecker::TypecheckFrom;
use crate::util;

use super::annotation::{AnnotationSpec, AnnotationTarget};
use super::{Scope, Type, Typed};
use codemap::Span;
use codemap_diagnostic::{Diagnostic, Level, SpanLabel, SpanStyle};
use payas_model::model::mapped_arena::MappedArena;
use serde::{Deserialize, Serialize, Serializer};

#[derive(Serialize, Deserialize, Debug, Clone, PartialEq)]
pub struct AnnotationMap {
    #[serde(serialize_with = "ordered_map")] // serialize with ordered_map to sort by key
    annotations: HashMap<String, AstAnnotation<Typed>>,

    /// Spans of the annotations (also keeps track of duplicate annotations).
    #[serde(skip_serializing)]
    #[serde(skip_deserializing)]
    spans: HashMap<String, Vec<Span>>,
}

fn ordered_map<S: Serializer>(
    value: &HashMap<String, AstAnnotation<Typed>>,
    serializer: S,
) -> Result<S::Ok, S::Error> {
    let ordered = value.iter().collect::<BTreeMap<_, _>>();
    ordered.serialize(serializer)
}

impl AnnotationMap {
    pub fn new(ast_annotations: &[AstAnnotation<Untyped>]) -> Self {
        let mut annotations = HashMap::new();
        let mut spans: HashMap<String, Vec<Span>> = HashMap::new();

        for a in ast_annotations {
            match spans.get_mut(&a.name) {
                Some(spans) => spans.push(a.span),
                None => {
                    annotations.insert(a.name.clone(), AstAnnotation::shallow(a));
                    spans.insert(a.name.clone(), vec![a.span]);
                }
            }
        }

        AnnotationMap { annotations, spans }
    }

    pub fn contains(&self, name: &str) -> bool {
        self.annotations.contains_key(name)
    }

    pub fn get(&self, name: &str) -> Option<&AstAnnotationParams<Typed>> {
        self.annotations.get(name).map(|a| &a.params)
    }

    pub fn pass(
        &mut self,
        target: AnnotationTarget,
        type_env: &MappedArena<Type>,
        annotation_env: &HashMap<String, AnnotationSpec>,
        scope: &Scope,
        errors: &mut Vec<Diagnostic>,
    ) -> bool {
        for (name, spans) in &self.spans {
            if spans.len() > 1 {
                let mut span_labels = vec![SpanLabel {
                    span: spans[0],
                    label: Some("previously defined here".to_owned()),
                    style: SpanStyle::Secondary,
                }];

                for span in &spans[1..] {
                    span_labels.push(SpanLabel {
                        span: *span,
                        label: Some("redefined here".to_owned()),
                        style: SpanStyle::Primary,
                    });
                }

                errors.push(Diagnostic {
                    level: Level::Error,
                    message: format!("Duplicate definitions of annotation `{}`", name),
                    code: Some("A000".to_string()),
                    spans: span_labels,
                });
            }
        }

        let mut changed = false;
        for annotation in self.annotations.values_mut() {
            match annotation_env.get(&annotation.name) {
                Some(spec) => {
                    let targets = spec.targets;
                    if !targets.contains(&target) {
                        let targets_str = util::join_strings(
                            &targets
                                .iter()
                                .map(|t| format!("{:?}", t).to_lowercase())
                                .collect::<Vec<_>>(),
                            Some("or"),
                        );

                        errors.push(Diagnostic {
                            level: Level::Error,
                            message: format!("Invalid target for annotation `{}`", annotation.name),
                            code: Some("A000".to_string()),
                            spans: vec![SpanLabel {
                                span: annotation.span,
                                label: Some(format!("only applies to targets: {}", targets_str)),
                                style: SpanStyle::Primary,
                            }],
                        });
                    }

                    let annot_changed = annotation.pass(type_env, annotation_env, scope, errors);
                    changed |= annot_changed;
                }
                None => {
                    errors.push(Diagnostic {
                        level: Level::Error,
                        message: format!("Unknown annotation `{}`", annotation.name),
                        code: Some("A000".to_string()),
                        spans: vec![SpanLabel {
                            span: annotation.span,
                            label: None,
                            style: SpanStyle::Primary,
                        }],
                    });
                    return false;
                }
            }
        }

        changed
    }
}
