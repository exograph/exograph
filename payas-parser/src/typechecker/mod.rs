mod annotation;
mod annotation_map;
mod annotation_params;
mod expression;
mod field;
mod field_type;
mod logical_op;
mod model;
mod relational_op;
mod selection;
mod service;
mod typ;

use std::collections::HashMap;

use anyhow::{anyhow, Result};
use codemap::CodeMap;
use codemap_diagnostic::{ColorConfig, Diagnostic, Emitter};
use serde::{Deserialize, Serialize};

pub(super) use annotation_map::AnnotationMap;

pub(super) use typ::{PrimitiveType, Type};

use crate::ast::ast_types::{AstModel, AstService, NodeTypedness};
use crate::ast::ast_types::{AstSystem, Untyped};
use payas_model::model::mapped_arena::MappedArena;

use self::annotation::{AnnotationSpec, AnnotationTarget, MappedAnnotationParamSpec};

pub struct Scope {
    pub enclosing_model: Option<String>,
}

#[derive(Serialize, Deserialize, Debug, Clone, PartialEq)]
pub struct Typed;
impl NodeTypedness for Typed {
    type FieldSelection = Type;
    type RelationalOp = Type;
    type Expr = Type;
    type LogicalOp = Type;
    type Field = Type;
    type Annotations = AnnotationMap;
    type Type = bool;
}

pub trait TypecheckInto<T> {
    fn shallow(&self) -> T;
    fn pass(
        &self,
        typ: &mut T,
        type_env: &MappedArena<Type>,
        annotation_env: &HashMap<String, AnnotationSpec>,
        scope: &Scope,
        errors: &mut Vec<Diagnostic>,
    ) -> bool;
}

pub trait TypecheckFrom<T>
where
    Self: Sized,
{
    fn shallow(untyped: &T) -> Self;
    fn pass(
        &mut self,
        type_env: &MappedArena<Type>,
        annotation_env: &HashMap<String, AnnotationSpec>,
        scope: &Scope,
        errors: &mut Vec<Diagnostic>,
    ) -> bool;
}

fn populate_type_env(env: &mut MappedArena<Type>) {
    // TODO: maybe we don't need to do this manually
    env.add("Boolean", Type::Primitive(PrimitiveType::Boolean));
    env.add("Int", Type::Primitive(PrimitiveType::Int));
    env.add("Float", Type::Primitive(PrimitiveType::Float));
    env.add("Decimal", Type::Primitive(PrimitiveType::Decimal));
    env.add("String", Type::Primitive(PrimitiveType::String));
    env.add("LocalTime", Type::Primitive(PrimitiveType::LocalTime));
    env.add(
        "LocalDateTime",
        Type::Primitive(PrimitiveType::LocalDateTime),
    );
    env.add("LocalDate", Type::Primitive(PrimitiveType::LocalDate));
    env.add("Instant", Type::Primitive(PrimitiveType::Instant));
    env.add("Json", Type::Primitive(PrimitiveType::Json));

    env.add("Claytip", Type::Primitive(PrimitiveType::ClaytipInjected));

    env.add(
        "Operation",
        Type::Primitive(PrimitiveType::Interception("Operation".to_string())),
    );
}

fn populate_annotation_env(env: &mut HashMap<String, AnnotationSpec>) {
    let annotations = [
        (
            "access",
            AnnotationSpec {
                targets: &[
                    AnnotationTarget::Model,
                    AnnotationTarget::Field,
                    AnnotationTarget::Method,
                ],
                no_params: false,
                single_params: true,
                mapped_params: Some(&[
                    MappedAnnotationParamSpec {
                        name: "query",
                        optional: true,
                    },
                    MappedAnnotationParamSpec {
                        name: "mutation",
                        optional: true,
                    },
                    MappedAnnotationParamSpec {
                        name: "create",
                        optional: true,
                    },
                    MappedAnnotationParamSpec {
                        name: "update",
                        optional: true,
                    },
                    MappedAnnotationParamSpec {
                        name: "delete",
                        optional: true,
                    },
                ]),
            },
        ),
        (
            "autoincrement",
            AnnotationSpec {
                targets: &[AnnotationTarget::Field],
                no_params: true,
                single_params: false,
                mapped_params: None,
            },
        ),
        (
            "bits",
            AnnotationSpec {
                targets: &[AnnotationTarget::Field],
                no_params: false,
                single_params: true,
                mapped_params: None,
            },
        ),
        (
            "column",
            AnnotationSpec {
                targets: &[AnnotationTarget::Field],
                no_params: false,
                single_params: true,
                mapped_params: None,
            },
        ),
        (
            "dbtype",
            AnnotationSpec {
                targets: &[AnnotationTarget::Field],
                no_params: false,
                single_params: true,
                mapped_params: None,
            },
        ),
        (
            "external",
            AnnotationSpec {
                targets: &[AnnotationTarget::Service],
                no_params: false,
                single_params: true,
                mapped_params: None,
            },
        ),
        (
            "before",
            AnnotationSpec {
                targets: &[AnnotationTarget::Interceptor],
                no_params: false,
                single_params: true,
                mapped_params: None,
            },
        ),
        (
            "after",
            AnnotationSpec {
                targets: &[AnnotationTarget::Interceptor],
                no_params: false,
                single_params: true,
                mapped_params: None,
            },
        ),
        (
            "around",
            AnnotationSpec {
                targets: &[AnnotationTarget::Interceptor],
                no_params: false,
                single_params: true,
                mapped_params: None,
            },
        ),
        (
            "inject",
            AnnotationSpec {
                targets: &[AnnotationTarget::Argument],
                no_params: true,
                single_params: true,
                mapped_params: None,
            },
        ),
        (
            "jwt",
            AnnotationSpec {
                targets: &[AnnotationTarget::Field],
                no_params: true,
                single_params: true,
                mapped_params: None,
            },
        ),
        (
            "length",
            AnnotationSpec {
                targets: &[AnnotationTarget::Field],
                no_params: false,
                single_params: true,
                mapped_params: None,
            },
        ),
        (
            "pk",
            AnnotationSpec {
                targets: &[AnnotationTarget::Field],
                no_params: true,
                single_params: false,
                mapped_params: None,
            },
        ),
        (
            "plural_name",
            AnnotationSpec {
                targets: &[AnnotationTarget::Model],
                no_params: false,
                single_params: true,
                mapped_params: None,
            },
        ),
        (
            "precision",
            AnnotationSpec {
                targets: &[AnnotationTarget::Field],
                no_params: false,
                single_params: true,
                mapped_params: None,
            },
        ),
        (
            "range",
            AnnotationSpec {
                targets: &[AnnotationTarget::Field],
                no_params: false,
                single_params: false,
                mapped_params: Some(&[
                    MappedAnnotationParamSpec {
                        name: "min",
                        optional: false,
                    },
                    MappedAnnotationParamSpec {
                        name: "max",
                        optional: false,
                    },
                ]),
            },
        ),
        (
            "scale",
            AnnotationSpec {
                targets: &[AnnotationTarget::Field],
                no_params: false,
                single_params: true,
                mapped_params: None,
            },
        ),
        (
            "size",
            AnnotationSpec {
                targets: &[AnnotationTarget::Field],
                no_params: false,
                single_params: true,
                mapped_params: None,
            },
        ),
        (
            "table",
            AnnotationSpec {
                targets: &[AnnotationTarget::Model],
                no_params: false,
                single_params: true,
                mapped_params: None,
            },
        ),
    ];

    for (name, spec) in annotations {
        env.insert(name.to_owned(), spec);
    }
}

pub fn build(ast_system: AstSystem<Untyped>, codemap: CodeMap) -> Result<MappedArena<Type>> {
    let mut ast_service_models: Vec<AstModel<Untyped>> = vec![];

    let mut types_arena: MappedArena<Type> = MappedArena::default();
    let mut annotation_env = HashMap::new();
    populate_type_env(&mut types_arena);
    populate_annotation_env(&mut annotation_env);

    let mut emitter = Emitter::stderr(ColorConfig::Always, Some(&codemap));

    for service in ast_system.services.iter() {
        ast_service_models.extend(service.models.clone());
        types_arena.add(&service.name, Type::Service(AstService::shallow(service)));
    }

    let ast_types = [ast_system.models.as_slice(), ast_service_models.as_slice()].concat();
    let ast_services = ast_system.services;

    for model in ast_types.iter() {
        types_arena.add(
            model.name.as_str(),
            Type::Composite(AstModel::shallow(model)),
        );
    }

    loop {
        let mut did_change = false;
        let init_scope = Scope {
            enclosing_model: None,
        };

        let mut errors = Vec::new();

        for model in ast_types.iter() {
            let mut typ = types_arena.get_by_key(model.name.as_str()).unwrap().clone();
            if let Type::Composite(c) = &mut typ {
                let pass_res = c.pass(&types_arena, &annotation_env, &init_scope, &mut errors);
                if pass_res {
                    *types_arena.get_by_key_mut(model.name.as_str()).unwrap() = typ;
                    did_change = true;
                }
            } else {
                panic!()
            }
        }

        for service in ast_services.iter() {
            let mut typ = types_arena
                .get_by_key(service.name.as_str())
                .unwrap()
                .clone();
            if let Type::Service(s) = &mut typ {
                let pass_res = s.pass(&types_arena, &annotation_env, &init_scope, &mut errors);
                if pass_res {
                    *types_arena.get_by_key_mut(service.name.as_str()).unwrap() = typ;
                    did_change = true;
                }
            } else {
                panic!()
            }
        }

        if !did_change {
            if !errors.is_empty() {
                emitter.emit(&errors);
                return Err(anyhow!("Could not process input clay files"));
            } else {
                return Ok(types_arena);
            }
        }
    }
}

#[cfg(test)]
pub mod test_support {
    use super::*;
    use crate::parser::*;

    pub fn build(src: &str) -> Result<MappedArena<Type>> {
        let (parsed, codemap) = parse_str(src);
        super::build(parsed, codemap)
    }

    pub fn parse_sorted(src: &str) -> Vec<(String, Type)> {
        let checked = build(src).unwrap();

        let mut entries: Vec<_> = checked
            .keys()
            .map(|key| (key.clone(), checked.get_by_key(key).unwrap().clone()))
            .collect();

        entries.sort_by_key(|pair| pair.0.to_owned());
        entries
    }
}

#[cfg(test)]
mod tests {
    use super::test_support::*;

    #[test]
    fn simple() {
        let src = r#"
        model User {
          doc: Doc @column("custom_column") @access(self.role == "role_admin" || self.role == "role_superuser" || self.doc.is_public)
          role: String
        }

        model Doc {
          is_public: Boolean
        }
        "#;

        assert_typechecking(src);
    }

    #[test]
    fn with_auth_context_use_in_field_annotation() {
        let src = r#"
        context AuthContext {
            role: String @jwt
        }
    
        model Doc {
          is_public: Boolean 
          content: String @access(AuthContext.role == "ROLE_ADMIN" || self.is_public)
        }
        "#;

        assert_typechecking(src);
    }

    #[test]
    fn with_auth_context_use_in_type_annotation() {
        let src = r#"
        context AuthContext {
            role: String @jwt
        }
    
        @access(AuthContext.role == "ROLE_ADMIN" || self.is_public)
        model Doc {
          is_public: Boolean 
          content: String 
        }
        "#;

        assert_typechecking(src);
    }

    #[test]
    fn insignificant_whitespace() {
        let typical = r#"
        @table("venues")
        model Venue {
            id: Int @column("idx") @pk
            name: String
        }
        "#;

        let with_whitespace = r#"

        @table ( "venues" )
        model    Venue   
        {
            id:   Int   @column(  "idx"  )    
            @pk 
            
            name:String

        }

        "#;

        let typical_parsed = serde_json::to_string(&parse_sorted(typical)).unwrap();
        let with_whitespace_parsed = serde_json::to_string(&parse_sorted(with_whitespace)).unwrap();
        assert_eq!(typical_parsed, with_whitespace_parsed);
    }

    #[test]
    fn unknown_annotation() {
        let src = r#"
        @asdf
        model User {
        }
        "#;

        assert_err(src);
    }

    fn duplicate_annotation() {
        let src = r#"
        @table("users")
        @table("users")
        model User {
        }
        "#;

        assert_err(src);
    }

    fn invalid_annotation_parameter_type() {
        let expected_none = r#"
        model User {
            id: Int @pk("asdf")
        }
        "#;

        let expected_single = r#"
        @table
        model User {
        }
        "#;

        let expected_map = r#"
        model User {
            id: Int @range(5)
        }
        "#;

        assert_err(expected_none);
        assert_err(expected_single);
        assert_err(expected_map);
    }

    fn duplicate_annotation_mapped_param() {
        let src = r#"
        model User {
            id: Int @range(min=5, max=10, min=3)
        }
        "#;

        assert_err(src);
    }

    fn unknown_annotation_mapped_param() {
        let src = r#"
        model User {
            id: Int @range(min=5, maxx=10)
        }
        "#;

        assert_err(src);
    }

    fn invalid_annotation_target() {
        let model = r#"
        @pk
        model User {
        }
        "#;

        let field = r#"
        model User {
            id: Int @table("asdf")
        }
        "#;

        assert_err(model);
        assert_err(field);
    }

    fn assert_typechecking(src: &str) {
        insta::with_settings!({sort_maps => true}, {
            insta::assert_yaml_snapshot!(build(src).unwrap())
        });
    }

    fn assert_err(src: &str) {
        assert!(build(src).is_err());
    }
}
