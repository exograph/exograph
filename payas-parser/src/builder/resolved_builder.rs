use payas_model::model::mapped_arena::MappedArena;

use crate::typechecker::{CompositeType, PrimitiveType, Type, TypedField};
use serde::{Deserialize, Serialize};

#[derive(Serialize, Deserialize, Debug, Clone, PartialEq)]
pub enum ResolvedType {
    Primitive(PrimitiveType),
    Composite(ResolvedCompositeType),
}

#[derive(Serialize, Deserialize, Debug, Clone, PartialEq)]
pub struct ResolvedCompositeType {
    pub name: String,
    pub fields: Vec<ResolvedField>,
    pub table_name: String,
}

#[derive(Serialize, Deserialize, Debug, Clone, PartialEq)]
pub struct ResolvedField {
    pub name: String,
    pub typ: ResolvedFieldType,
    pub column_name: String,
    pub is_pk: bool,
    pub is_autoincrement: bool,
}

#[derive(Serialize, Deserialize, Debug, Clone, PartialEq)]
pub enum ResolvedFieldType {
    Plain(String), // Should really be Id<ResolvedType>, but using String since the former is not serializable as needed by the insta crate
    Optional(Box<ResolvedFieldType>),
    List(Box<ResolvedFieldType>),
}

impl ResolvedCompositeType {
    pub fn pk_field(&self) -> Option<&ResolvedField> {
        self.fields.iter().find(|f| f.is_pk)
    }
}

impl ResolvedType {
    fn name(&self) -> &str {
        match self {
            ResolvedType::Primitive(pt) => pt.name(),
            ResolvedType::Composite(ResolvedCompositeType { name, .. }) => name,
        }
    }

    pub fn as_primitive(&self) -> PrimitiveType {
        match &self {
            ResolvedType::Primitive(p) => p.clone(),
            _ => panic!("Not a primitive: {:?}", self),
        }
    }

    // useful for relation creation
    pub fn as_composite<'a>(&'a self) -> &'a ResolvedCompositeType {
        match &self {
            ResolvedType::Composite(c) => c,
            _ => panic!("Cannot get inner composite of type {:?}", self),
        }
    }
}

impl ResolvedFieldType {
    pub fn deref<'a>(&'a self, types: &'a MappedArena<ResolvedType>) -> &'a ResolvedType {
        match self {
            ResolvedFieldType::Plain(name) => types.get_by_key(&name).unwrap(),
            ResolvedFieldType::Optional(underlying) | ResolvedFieldType::List(underlying) => {
                underlying.deref(types)
            }
        }
    }
}

/// Consume typed-checked types and build resolved types
/// Resolved types normalized annotations in that if an annotaion isn't provided,
/// this build compute the default value and sets that information in the resulting resolved type
pub fn build(types: MappedArena<Type>) -> MappedArena<ResolvedType> {
    let mut resolved_types = build_shallow(&types);
    build_expanded(types, &mut resolved_types);
    resolved_types
}

fn build_shallow(types: &MappedArena<Type>) -> MappedArena<ResolvedType> {
    let mut resolved_arena: MappedArena<ResolvedType> = MappedArena::default();

    for (_, typ) in types.iter() {
        let resolved = match typ {
            Type::Primitive(pt) => ResolvedType::Primitive(pt.clone()),
            Type::Composite(ct) => {
                let table_name = ct
                    .get_annotation("table")
                    .map(|a| a.params[0].as_string())
                    .unwrap_or_else(|| ct.name.clone());
                ResolvedType::Composite(ResolvedCompositeType {
                    name: ct.name.clone(),
                    fields: vec![],
                    table_name,
                })
            }
            _ => panic!(""),
        };

        let name = resolved.name().to_string();

        resolved_arena.add(&name, resolved);
    }

    resolved_arena
}

fn build_expanded(types: MappedArena<Type>, resolved_types: &mut MappedArena<ResolvedType>) {
    for (_, typ) in types.iter() {
        if let Type::Composite(ct) = typ {
            let existing_type_id = resolved_types.get_id(&ct.name).unwrap();
            let existing_type = &resolved_types[existing_type_id];
            let resolved_fields = ct
                .fields
                .iter()
                .map(|field| ResolvedField {
                    name: field.name.clone(),
                    typ: resolve_field_type(&field.typ, &types, resolved_types),
                    column_name: compute_column_name(ct, field, &types),
                    is_pk: field.get_annotation("pk").is_some(),
                    is_autoincrement: field.get_annotation("autoincrement").is_some(),
                })
                .collect();
            if let ResolvedType::Composite(ResolvedCompositeType {
                name, table_name, ..
            }) = existing_type
            {
                let expanded = ResolvedType::Composite(ResolvedCompositeType {
                    name: name.clone(),
                    fields: resolved_fields,
                    table_name: table_name.clone(),
                });
                resolved_types[existing_type_id] = expanded;
            }
        }
    }
}

fn compute_column_name(
    enclosing_type: &CompositeType,
    field: &TypedField,
    types: &MappedArena<Type>,
) -> String {
    fn default_column_name(
        enclosing_type: &CompositeType,
        field: &TypedField,
        types: &MappedArena<Type>,
    ) -> String {
        match &field.typ {
            Type::Optional(_) => field.name.to_string(),
            Type::List(_) => format!("{}_id", enclosing_type.name.to_ascii_lowercase()),
            Type::Reference(type_name) => {
                let field_type = types.get_by_key(&type_name).unwrap();
                match field_type {
                    Type::Composite(_) => format!("{}_id", field.name),
                    _ => field.name.clone(),
                }
            }
            _ => panic!(""),
        }
    }

    field
        .get_annotation("column")
        .map(|a| a.params[0].as_string())
        .unwrap_or_else(|| default_column_name(enclosing_type, field, types))
}

fn resolve_field_type(
    typ: &Type,
    types: &MappedArena<Type>,
    resolved_types: &MappedArena<ResolvedType>,
) -> ResolvedFieldType {
    match typ {
        Type::Optional(underlying) => ResolvedFieldType::Optional(Box::new(resolve_field_type(
            underlying,
            types,
            resolved_types,
        ))),
        Type::List(underlying) => ResolvedFieldType::List(Box::new(resolve_field_type(
            underlying,
            types,
            resolved_types,
        ))),
        Type::Reference(name) => ResolvedFieldType::Plain(name.to_owned()),
        t => panic!("Invalid type {:?}", t),
    }
}

#[cfg(test)]
mod tests {
    use payas_model::model::mapped_arena::sorted_values;

    use super::*;
    use crate::{parser, typechecker};

    #[test]
    fn with_annotations() {
        let src = r#"
        @table("custom_concerts")
        model Concert {
          id: Int @pk @autoincrement @column("custom_id")
          title: String @column("custom_title")
          venue: Venue @column("custom_venue_id")
        }
        
        @table("venues")
        model Venue {
          id: Int @pk @autoincrement @column("custom_id")
          name: String @column("custom_name")
          concerts: [Concert] @column("custom_venueid")
        }        
        "#;

        let parsed = parser::parse_str(src);
        let types = typechecker::build(parsed);

        let resolved = build(types);

        insta::assert_yaml_snapshot!(sorted_values(&resolved));
    }

    #[test]
    fn with_defaults() {
        // Note the swapped order between @pk and @autoincrement to assert that our parsing logic permits any order
        let src = r#"
        model Concert {
          id: Int @pk @autoincrement 
          title: String 
          venue: Venue 
        }
        
        model Venue             {
          id: Int  @autoincrement @pk 
          name:String 
          concerts: [Concert] 
        }        
        "#;

        let parsed = parser::parse_str(src);
        let types = typechecker::build(parsed);

        let resolved = build(types);

        insta::assert_yaml_snapshot!(sorted_values(&resolved));
    }
}
