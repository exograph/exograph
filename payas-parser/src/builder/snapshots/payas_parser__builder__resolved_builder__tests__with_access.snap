---
source: payas-parser/src/builder/resolved_builder.rs
expression: resolved
---
types:
  values:
    - - ~
      - Primitive: Boolean
    - - ~
      - Primitive: Int
    - - ~
      - Primitive: Float
    - - ~
      - Primitive: Decimal
    - - ~
      - Primitive: String
    - - ~
      - Primitive: LocalTime
    - - ~
      - Primitive: LocalDateTime
    - - ~
      - Primitive: LocalDate
    - - ~
      - Primitive: Instant
    - - ~
      - Primitive: Json
    - - ~
      - Primitive: Blob
    - - ~
      - Primitive: Uuid
    - - ~
      - Primitive: Claytip
    - - ~
      - Primitive: ClaytipPriv
    - - ~
      - Primitive:
          Interception: Operation
    - - ~
      - Composite:
          name: Concert
          plural_name: Concerts
          fields:
            - name: id
              typ:
                Plain: Int
              kind:
                Persistent:
                  column_name: id
                  self_column: true
                  is_pk: true
                  type_hint: ~
                  unique_constraints: []
              default_value: Autoincrement
            - name: title
              typ:
                Plain: String
              kind:
                Persistent:
                  column_name: title
                  self_column: true
                  is_pk: false
                  type_hint: ~
                  unique_constraints: []
              default_value: ~
            - name: public
              typ:
                Plain: Boolean
              kind:
                Persistent:
                  column_name: public
                  self_column: true
                  is_pk: false
                  type_hint: ~
                  unique_constraints: []
              default_value: ~
          kind:
            Persistent:
              table_name: concerts
          access:
            creation:
              LogicalOp:
                Or:
                  - RelationalOp:
                      Eq:
                        - FieldSelection:
                            Select:
                              - Single:
                                  - - AuthContext
                                  - Reference:
                                      index: 16
                                      generation: ~
                              - - role
                              - Reference:
                                  index: 4
                                  generation: ~
                        - StringLiteral:
                            - ROLE_ADMIN
                        - Primitive: Boolean
                  - FieldSelection:
                      Select:
                        - Single:
                            - - self
                            - Reference:
                                index: 17
                                generation: ~
                        - - public
                        - Reference:
                            index: 0
                            generation: ~
                  - Primitive: Boolean
            read:
              LogicalOp:
                Or:
                  - RelationalOp:
                      Eq:
                        - FieldSelection:
                            Select:
                              - Single:
                                  - - AuthContext
                                  - Reference:
                                      index: 16
                                      generation: ~
                              - - role
                              - Reference:
                                  index: 4
                                  generation: ~
                        - StringLiteral:
                            - ROLE_ADMIN
                        - Primitive: Boolean
                  - FieldSelection:
                      Select:
                        - Single:
                            - - self
                            - Reference:
                                index: 17
                                generation: ~
                        - - public
                        - Reference:
                            index: 0
                            generation: ~
                  - Primitive: Boolean
            update:
              LogicalOp:
                Or:
                  - RelationalOp:
                      Eq:
                        - FieldSelection:
                            Select:
                              - Single:
                                  - - AuthContext
                                  - Reference:
                                      index: 16
                                      generation: ~
                              - - role
                              - Reference:
                                  index: 4
                                  generation: ~
                        - StringLiteral:
                            - ROLE_ADMIN
                        - Primitive: Boolean
                  - FieldSelection:
                      Select:
                        - Single:
                            - - self
                            - Reference:
                                index: 17
                                generation: ~
                        - - public
                        - Reference:
                            index: 0
                            generation: ~
                  - Primitive: Boolean
            delete:
              LogicalOp:
                Or:
                  - RelationalOp:
                      Eq:
                        - FieldSelection:
                            Select:
                              - Single:
                                  - - AuthContext
                                  - Reference:
                                      index: 16
                                      generation: ~
                              - - role
                              - Reference:
                                  index: 4
                                  generation: ~
                        - StringLiteral:
                            - ROLE_ADMIN
                        - Primitive: Boolean
                  - FieldSelection:
                      Select:
                        - Single:
                            - - self
                            - Reference:
                                index: 17
                                generation: ~
                        - - public
                        - Reference:
                            index: 0
                            generation: ~
                  - Primitive: Boolean
  map:
    Blob:
      index: 10
      generation: ~
    Boolean:
      index: 0
      generation: ~
    Claytip:
      index: 12
      generation: ~
    ClaytipPriv:
      index: 13
      generation: ~
    Concert:
      index: 15
      generation: ~
    Decimal:
      index: 3
      generation: ~
    Float:
      index: 2
      generation: ~
    Instant:
      index: 8
      generation: ~
    Int:
      index: 1
      generation: ~
    Json:
      index: 9
      generation: ~
    LocalDate:
      index: 7
      generation: ~
    LocalDateTime:
      index: 6
      generation: ~
    LocalTime:
      index: 5
      generation: ~
    Operation:
      index: 14
      generation: ~
    String:
      index: 4
      generation: ~
    Uuid:
      index: 11
      generation: ~
contexts:
  values:
    - - ~
      - name: AuthContext
        fields:
          - name: role
            typ:
              Plain: String
            source:
              annotation: jwt
              value: role
    - ~
    - ~
    - ~
  map:
    AuthContext:
      index: 0
      generation: ~
services:
  values:
    - - ~
      - name: Logger
        script:
          - 47
          - 47
          - 32
          - 100
          - 101
          - 110
          - 111
          - 45
          - 102
          - 109
          - 116
          - 45
          - 105
          - 103
          - 110
          - 111
          - 114
          - 101
          - 45
          - 102
          - 105
          - 108
          - 101
          - 10
          - 47
          - 47
          - 32
          - 100
          - 101
          - 110
          - 111
          - 45
          - 108
          - 105
          - 110
          - 116
          - 45
          - 105
          - 103
          - 110
          - 111
          - 114
          - 101
          - 45
          - 102
          - 105
          - 108
          - 101
          - 10
          - 47
          - 47
          - 32
          - 84
          - 104
          - 105
          - 115
          - 32
          - 99
          - 111
          - 100
          - 101
          - 32
          - 119
          - 97
          - 115
          - 32
          - 98
          - 117
          - 110
          - 100
          - 108
          - 101
          - 100
          - 32
          - 117
          - 115
          - 105
          - 110
          - 103
          - 32
          - 96
          - 100
          - 101
          - 110
          - 111
          - 32
          - 98
          - 117
          - 110
          - 100
          - 108
          - 101
          - 96
          - 32
          - 97
          - 110
          - 100
          - 32
          - 105
          - 116
          - 39
          - 115
          - 32
          - 110
          - 111
          - 116
          - 32
          - 114
          - 101
          - 99
          - 111
          - 109
          - 109
          - 101
          - 110
          - 100
          - 101
          - 100
          - 32
          - 116
          - 111
          - 32
          - 101
          - 100
          - 105
          - 116
          - 32
          - 105
          - 116
          - 32
          - 109
          - 97
          - 110
          - 117
          - 97
          - 108
          - 108
          - 121
          - 10
          - 10
          - 10
        script_path: logger.js
        methods:
          - name: log
            operation_kind: Query
            is_exported: true
            access:
              creation:
                RelationalOp:
                  Eq:
                    - FieldSelection:
                        Select:
                          - Single:
                              - - AuthContext
                              - Reference:
                                  index: 16
                                  generation: ~
                          - - role
                          - Reference:
                              index: 4
                              generation: ~
                    - StringLiteral:
                        - ROLE_ADMIN
                    - Primitive: Boolean
              read:
                RelationalOp:
                  Eq:
                    - FieldSelection:
                        Select:
                          - Single:
                              - - AuthContext
                              - Reference:
                                  index: 16
                                  generation: ~
                          - - role
                          - Reference:
                              index: 4
                              generation: ~
                    - StringLiteral:
                        - ROLE_ADMIN
                    - Primitive: Boolean
              update:
                RelationalOp:
                  Eq:
                    - FieldSelection:
                        Select:
                          - Single:
                              - - AuthContext
                              - Reference:
                                  index: 16
                                  generation: ~
                          - - role
                          - Reference:
                              index: 4
                              generation: ~
                    - StringLiteral:
                        - ROLE_ADMIN
                    - Primitive: Boolean
              delete:
                RelationalOp:
                  Eq:
                    - FieldSelection:
                        Select:
                          - Single:
                              - - AuthContext
                              - Reference:
                                  index: 16
                                  generation: ~
                          - - role
                          - Reference:
                              index: 4
                              generation: ~
                    - StringLiteral:
                        - ROLE_ADMIN
                    - Primitive: Boolean
            arguments:
              - name: claytip
                typ:
                  Plain: Claytip
                is_injected: true
            return_type:
              Plain: Boolean
        interceptors: []
    - ~
    - ~
    - ~
  map:
    Logger:
      index: 0
      generation: ~

